### PR TITLE
Fix manual balance may return TaskStale error

### PR DESCRIPTION
**Pull request by yah01**
the balancing segments may be compacted, which causes TaskStale

Signed-off-by: yah01 <yang.cen@zilliz.com>

related #19386

### CODE REVIEW

**Comment by yah01** [comment]
rerun ut

File: internal/querycoordv2/handlers.go [hunk 1]
```diff
@@ -147,7 +148,32 @@ func (s *Server) balanceSegments(ctx context.Context, req *querypb.LoadBalanceRe
...
+	var err error
+	go func() {
+		for _, t := range tasks {
+			err = t.Wait()
+			if err != nil && !errors.Is(err, task.ErrTaskStale) {
+				cancel()
+				break
+			}
+		}
+		cancel()
```
Comment (congqixia) [inline internal/querycoordv2/handlers.go:172]: Does tasks have order here? Looks equivalent to a plain for-loop without goroutine here

Comment (yah01) [reply internal/querycoordv2/handlers.go:172]: No, the goroutine here makes the method able to return if the request's context canceled or timeout

File: internal/querycoordv2/handlers.go [hunk 2]
```diff
@@ -147,7 +148,32 @@ func (s *Server) balanceSegments(ctx context.Context, req *querypb.LoadBalanceRe
...
-	return task.Wait(ctx, Params.QueryCoordCfg.SegmentTaskTimeout, tasks...)
+	return waitBalanceTasks(ctx, Params.QueryCoordCfg.SegmentTaskTimeout, tasks...)
+}
+
+// waitBalanceTasks blocks until the tasks finished,
+// return the first returned err of task,
+// NOTE: this ignores the TaskStale error,
+// because the balancing segments may be compacted,
+// which should be ignored, more details #19386
+func waitBalanceTasks(ctx context.Context, timeout time.Duration, tasks ...task.Task) error {
```
Comment (congqixia) [inline internal/querycoordv2/handlers.go:159]: Maybe `waitBalanceTasks` shall be in `task/utils.go` with `Wait` function.
Or we could change the `task.Wait` function add error check method?

**Comment by yah01** [comment]
We found the manual balance with no specific segments will balance all segments from source node to destination node, even which are not in the given collection, and #20738 fixed this. 
The TaskStale error still exists when the balancing segments are compacted, but will happen with very small probability.

**Pull request closed without merging**